python/trezorlib/eos.py

```diff
@@ -0,0 +1,353 @@
+    for account in data['accounts']:
+        accounts.append(
+            messages.EosAuthorizationAccount(
+                account=messages.EosPermissionLevel(
+                    actor=name_to_number(account['permission']['actor']),
+                    permission=name_to_number(account['permission']['permission'])
+                ),
+                weight=int(account['weight'])
+            )
+        )
+
+    waits = []
+    for wait in data['waits']:
+        waits.append(
+            messages.EosAuthorizationWait(
+                wait_sec=int(wait['wait_sec']),
+                weight=int(wait['weight'])
+            )
+        )
+
+    return messages.EosAuthorization(
+        threshold=int(data['threshold']),
+        keys=keys,
+        accounts=accounts,
+        waits=waits,
+    )
+
+
+def parse_new_account(data):
+    owner = parse_authorization(data['owner'])
+    active = parse_authorization(data['active'])
+
+    return messages.EosActionNewAccount(
+        creator=name_to_number(data['creator']),
+        name=name_to_number(data['name']),
+        owner=owner,
+        active=active
+    )
+
+
+def parse_unknown(data):
+    data_bytes = bytes.fromhex(data)
+    return messages.EosActionUnknown(
+        data_size=len(data_bytes),
+        data_chunk=data_bytes
+    )
+
+
+def parse_action(action):
+    txAction = messages.EosTxActionAck()
+    data = action['data']
+
+    txAction.common = parse_common(action)
+
+    if action['account'] == 'eosio':
+        if action['name'] == 'voteproducer':
+            txAction.vote_producer = parse_vote_producer(data)
+        elif action['name'] == 'buyram':
+            txAction.buy_ram = parse_buy_ram(data)
+        elif action['name'] == 'buyrambytes':
+            txAction.buy_ram_bytes = parse_buy_rambytes(data)
+        elif action['name'] == 'sellram':
+            txAction.sell_ram = parse_sell_ram(data)
+        elif action['name'] == 'delegatebw':
+            txAction.delegate = parse_delegate(data)
+        elif action['name'] == 'undelegatebw':
+            txAction.undelegate = parse_undelegate(data)
+        elif action['name'] == 'refund':
+            txAction.refund = parse_refund(data)
+        elif action['name'] == 'updateauth':
+            txAction.update_auth = parse_updateauth(data)
+        elif action['name'] == 'deleteauth':
+            txAction.delete_auth = parse_deleteauth(data)
+        elif action['name'] == 'linkauth':
+            txAction.link_auth = parse_linkauth(data)
+        elif action['name'] == 'unlinkauth':
+            txAction.unlink_auth = parse_unlinkauth(data)
+        elif action['name'] == 'newaccount':
+            txAction.new_account = parse_new_account(data)
+    elif action['name'] == 'transfer':
+        txAction.transfer = parse_transfer(data)
+    else:
+        txAction.unknown = parse_unknown(data)
+
+    return txAction
+
+
+def parse_transaction_json(json):
+    tx = type('Transaction', (object,), {})()
```

Comment:
don't do this

this function should probably return `header, actions`, where `header` is `EosTxHeader` object, and `actions = [parse_action(a) for a in body['actions']]`

The `chain_id` parameter is strange. It seems that your source JSON only has two top-level keys, `chain_id` and `transaction`? In that case, `chain_id` should probably be a separate argument to `sign_tx`, and the passed `transaction` should just be `json['transaction']`.

(this presumes that "chain-id" is something like fork selection and is not part of the transaction data itself?)

Reply:
Restructured code. Done.